### ISSUE TITLE
Merge and Moves are on per user basis

### ISSUE THREAD

**Issue by jacksbox** [nextcloud/recognize#948]
### Which version of recognize are you using?

3.4.1

### Enabled Modes

Object recognition, Face recognition

### TensorFlow mode

WASM mode

### Downstream App

Files App

### Which Nextcloud version do you have installed?

27.0.0

### Which Operating system do you have installed?

Ubuntu 22.04

### Which database are you running Nextcloud on?

postgres

### Which Docker container are you using to run Nextcloud? (if applicable)

aio

### How much RAM does your server have?

8GiB

### What processor Architecture does your CPU have?

arm64

### Describe the Bug

Not sure if it's a bug or a feature.
I have two accounts which have access to a lot of shared pictures.
The recognition works quite well and there are just a few Presons which needs to be merged.
But if User A merges two Persons, or moves certain Faces to a Person, this only is applied for this user.
If User B Logs in, the Persons are in the previouse, unmerged state.
Is that intended? 

### Expected Behavior

Merged Persons and Moved Images should be applied to all users which have access to this images.

### To Reproduce

Have 2 users
Share all Pictures
Recognise
Merge some Persons for User A
User B sees the unmerged Persons...

**Comment by marcelklehr** [comment]
Yes this is intended. At some point it may be possible to share "photos of person A" with another user, which is probably what you want, but not at the moment. see #881

**Comment by jacksbox** [comment]
hm, not exactly, no.
I'll try to explain:

* there are 1000 Images, which are shared between user 1 and 2 (both can access all)
* classification and clustering finished
* clustering resulted in 3 Clusters, of Person A, B, C
* Now User 1 merges Cluster A and B, as this clusters are in fact the same Person
* User 1 has now only 2 Clusters, A and C
* User 2 (with access to the same images) still has Clusters A, B and C

I would expected - better wish ;) - that User 2 now also just has clusters A and C to prevent that user from doing the merge as well (which is easy in this example, but when we talk 10000s of images, there will be a lot of misclassifications which can be corrected)

---

It seems that for now each user uses individual clusters which makes this hard to achieve.
But why not share clusters between all users? Show a user only pictures of a cluster to which the user has access to. If a cluster is empty for a user (because that user does not have access to any picture from it) it can be hidden / not shown.
But I guess that could result in more misclassifications for setups where the users do not have many overlapping persons? Or not? Sorry, just rambling about...

**Comment by marcelklehr** [comment]
mmmh. Interesting thought. Let me loop in @jancborchardt here to see what he thinks.

**Comment by Forza-tng** [comment]
I can see both arguments. However, my personal opinion is to have a choice/setting to share metadata (people, object) with specific groups or people. My use case is family photos which I share with my partner. We have 100k photos from two decades stored and shared on our Nextcloud. It is great if we could jointly work on tagging photos and share the results, as doing duplicate work isn't really feasible.

**Comment by rilcy75** [comment]
Hi,
Great job!
For me this feature is a must have. As many I use this NC app to share thousands of photo with my (huge) family and friend. Making recognizing and naming people cross NC users is a key features to address my objective.
It could avoid people to use additionnal platform to host dedicated solution for photo sharing.

**Comment by Olen** [comment]
So, a challenge here is that not only the clusters are per user, but the detections as well, even if the files are shared:

This is an example of an image with 3 faces, shared in a groupfolder with 3 members:
```
  id   | fileid  |                        path                     |  user_id   | cluster_id                     
-------+---------+-------------------------------------------------+------------+------------
 18465 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | XXXX       |       1163
 18468 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | XXXX       |         -1
 18462 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | XXXX       |        134
 18461 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | YYYY       |        136
 18464 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | YYYY       |        238
 18467 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | YYYY       |         55
 18469 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | ZZZZ       |         -1
 18463 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | ZZZZ       |        140
 18466 | 4695116 | __groupfolders/1/2014-12-24_184755-P1000037.JPG | ZZZZ       |          1
(9 rows)
```
Notice the same file_id and path, but different IDs in the face_detection table.

So each user has an entry in the table for each person (face) in each image.  And each of these entries have a unique cluster_id.
The person-identifier (face) is the x/y/vector, so there are some different aproaches possible:

* Remove `user_id` from the table `oc_recognize_face_clusters`
* Create a new table oc_recognize_face_clusters_users with two columns: `cluster_id` and `user_id`
* By default continue adding one user per cluster as today.
* Allow more user_ids to be added to the same cluster_id in the new table

This will need some additional code to update all the entries in the `oc_recognize_face_detections` table

1. Create a shared cluster
2. Add users A, B and C to the shared cluster
3. User A moves a "face" (actually combination of file_id, x/y/vector) to shared cluster
4. Magically - the same "face" (same combination of file_id, x/y/vector) is moved to the same shared cluster_id for user B and C

But we need to figure out how to handle different cases.
- What if User B moves the "face" from the shared cluster to a private cluster - or removes the face?
- What if User A moves a non-shared image to the shared cluster?
- What if a shared image file in a shared cluster is moved to a non-shared folder?




Another option is to create what I would call "synced" clusters.
- User A invites User B to sync cluster X
- When the invite is accepted a new cluster Y is created for User B
- All combinations of file_id+vector that exist in cluster X for User A is then duplicated for User B to cluster Y.
- Any further modifications User A does to Cluster X is replicated for User B in Cluster Y
- And any modifications User B does to Cluster Y is replicated to User A in Cluster X

This would need some additional columns in the oc_recognize_face_detections to add a timestamp or version id to ensure that the sync can be done properly, but is a slightly cleaner way as it can basically be done by an external script, and does not need much magic behind the scene.

Non-shared images are no problem here, because if the file_id is not found for User B, the image is just not added to User Bs synced cluster.

But it still have some challenges, like 
- What if User B moves a face from the synced cluster to a non-synced cluster.  Where should this image be placed for User A?
- There are possibilities for race-conditions where both users try to move a face to a different synced clusters at the same time. 

So there is no easy way to do this as far as I can see, and it will require some work to really get it functional.

**Comment by vbelloir** [comment]
Hello folks,
I give a try to recognize to see if it fits my needs.
I am in the same situation as other.
I used my self hosted nextcloud as a familly cloud, and I share 20 years of photo with my wife, and maybe later with childrens.

I give a try with a bunch of 1000 pictures. Face recognition is good, but with old photos, I have to help merging with already named cluster.
If I spend hours, to parse the whole collection, I don't want that my wife and childrens have to do the same.

In my opinion, this is really a must have feature.

**Comment by mkulich** [comment]
Just came here to say this would be a great feature. I have the same use case where I have 10's of thousands of old family photos shared between users and it would be great if everyone could collaborate on them tagging faces.

**Comment by rivendellstuff** [comment]
just realized after letting it scan for awhile that its only doing it for my username and not my family for the 100k plus photos we have. this makes the feature completely useless for my use case. damn.  at least tags persist. otherwise it would be completely useless to use nextcloud for photos at all

**Comment by Chrispletsch** [comment]
I´m facing the same use case.
It would be great to have an option to decide how to handle this topic.
For me it is the same as others before, me and my wife and familiy should not categorize faces two and more times without any additional benefit.
For me this is a must have feature.
On the otherhand memories ist just a great App!

**Comment by tobiaseigen** [comment]
For my use case this would be very helpful as well. My nextcloud is a family with just a few users, and we all access the same photos library.

**Comment by marcelklehr** [comment]
Please refrain from commenting without adding substantial information. If you wish to express support, use the emoji reactions, please.

**Comment by marcelklehr** [comment]
Note, that you do have a little influence over what Nextcloud GmbH works on: We don't promise anything, but every release cycle we try to work on enhancements that get a lot of upvotes, so you may express your support for this by giving this issue an upvote.

**Comment by jancborchardt** [comment]
@marcelklehr to me it sounds like a big possible privacy issue, much like the "shared tags" concept where private tags is something people want and need.

**Comment by marcelklehr** [comment]
Well, currently, even if you shared all your photos in nextcloud to the accounts of your family members, they will not see the same face clusters as you do and will have to do the same manual book keeping all over again to get clean face clusters. I guess the more pressing issue might be to improve automatic face clustering to obviate the need to do manual cleaning, still I think sharing faces is a legitimate use case.

**Comment by jancborchardt** [comment]
@marcelklehr right, so definitely UX-wise it’s nice, especially if it’s something that is done automatically. E.g. if you share some photos, also share the relevant face clusters. But if you set any custom names, they should of course not be shared?

That would improve the UX (since as you say, otherwise likely the same clusters would be found anyway), while not sharing the private info (the names you set manually).

**Comment by marcelklehr** [comment]
That's a nice idea! I like it. The only downside I see is that a users whom I have shared photos with can destroy the face clusters I have built.

**Comment by jerrac** [comment]
I came across this issue due to figuring out that my initial `recognize:classify` run was running the classification on each file for every user that has access to the team folder containing the file. I'm on day 2+ of that initial run... I am not looking forward to what happens when I add more than 3 accounts to that folder...

(Is it actually scanning per user? I think it is because the output of the command uses the full path to the file including the user id, not the __groupfolders path I see in oc_filecache.)

Like https://github.com/nextcloud/recognize/issues/948#issuecomment-1725959193 I want to have family photo's scanned by recognize, and then work as a team to get the data accurate. 

In addition, I really really don't want to be scanning a file more than one time per run of `recognize:classify`. 

From what I can tell in this issue, the primary concerns are: 

There could be issues if one account merges one cluster into another, but another account merges the same cluster into a different one. Or the equivalent for individual files.
  * For merges: if user A merges cluster 1 into cluster 2, would user B really have the opportunity to try and merge cluster 1 into cluster 3? I'd think the first merge would cause the second to fail with a "cluster 1 doesn't exist" error.
  * For moving a file face detection item between clusters, wouldn't it either error due to the item no longer being in the original cluster, or would it just ignore the move and put the item into the target cluster?

The "label" of a face/cluster might need to be kept private. This could be addressed by the changes I suggest below.

For files in team folders:
* Clusters are owned by the folder.
* If advanced permissions are enabled, then team managers can merge/rename/etc. clusters and items.
* If advanced permissions are not enabled, then all team members can merge/rename/etc. clusters and items.
* Since it is in a team folder, there should be no need to worry about labels being private.

For shared files/folders:
* Add custom permissions to the share dialog for recognize.
  * Recognize data permissions:
    * Recognize face/tags/etc. - rescan
    * Recognize face/tags/etc. - read only 
    * Recognize face/tags/etc. - edit
      * If the edit option is enabled, Recognize would copy the cluster data to a new cluster for that share.
    * Recognize face/tags/etc. - include labels
      * If not checked, then just the id is shared. This solves the issue of keeping things private.
        * If someone edits the label, then that label only applies to the shared cluster and is not private.
* Rescan means that instead of copying from the owner, the file is rescanned and that is what is set up for the share.
* Copy the "clusters"/"detection's" for the items to share specific versions.
* In read only mode: only the owner can modify clusters.
* There would need to be UI elements that lets the owner sync changes to the cluster to and from the clusters specific to the user.
* When a new file is added to a shared directory, the detection would be added to the owner first, then copied to the share.

In the UI for Photo's and Memories, I'd add a prefix to the various faces indicating if they are user, shared, or team folder. For shared and team folders, I'd add an indicator for which share/team folder the cluster is in. Then I'd make sure that you can't move a face detection between different user, share, or team folder clusters. 

So, to recap my thoughts: Split clusters up into user, share, and team folder types. This lets us adapt how everything works based on the type of the cluster. And that adaptability allows for better privacy, and fewer scans of the files.

Anyway, hopefully this adds to the discussion. Apologies if I didn't quite get the right terminology for things...
